File: desktop/frontend/src/lib/slug.test.ts
import { describe, it, expect } from "vitest";
import { slug, slugTyping, isSlug, displayName } from "./slug";

// These cases mirror internal/config/slug_test.go one for one. When either side
// changes, both must — the Go implementation is the authority and this file is
// what catches the drift.
describe("slug", () => {
  it.each([
    ["Nori App", "nori-app"],
    ["nori-app", "nori-app"],
    ["  Nori   App  ", "nori-app"],
    ["Okane", "okane"],
    ["my_project.v2", "my_project.v2"],
    ["Ünïcødé Ãpp", "n-c-d-pp"],
    ["日本語", ""],
    ["a/b", "a-b"],
    ["../etc", "etc"],
    ["...", ""],
    ["", ""],
    ["   ", ""],
    ["---", ""],
  ])("slug(%o) === %o", (input, want) => {
    expect(slug(input)).toBe(want);
  });

  it("is idempotent and self-consistent", () => {
    for (const input of ["Nori App", "a/b", "../etc", "my_project.v2", "Okane"]) {
      const s = slug(input);
      if (s === "") continue;
      expect(slug(s)).toBe(s);
      expect(isSlug(s)).toBe(true);
    }
  });

  it("never lets a path separator survive", () => {
    // The id becomes a directory name; a "/" would escape the segment entirely.
    for (const input of ["a/b", "../../etc/passwd", "a\\b"]) {
      expect(slug(input)).not.toMatch(/[/\\]/);
    }
  });
});

describe("isSlug", () => {
  it.each(["", "Okane", "Nori App", "a/b", "-lead", "trail-", "..", "."])(
    "rejects %o",
    (input) => {
      expect(isSlug(input)).toBe(false);
    },
  );
});

describe("slugTyping", () => {
  it("keeps a trailing separator so a hyphen can be typed at all", () => {
    expect(slugTyping("nori-")).toBe("nori-");
  });

  it("converges on the slug when typed one keystroke at a time", () => {
    let typed = "";
    for (const ch of "Nori App") typed = slugTyping(typed + ch);
    expect(typed).toBe("nori-app");
  });
});

describe("displayName", () => {
  it("falls back to the id when there is no label", () => {
    expect(displayName({ name: "nori-app" })).toBe("nori-app");
    expect(displayName({ name: "nori-app", label: "" })).toBe("nori-app");
    expect(displayName({ name: "nori-app", label: "   " })).toBe("nori-app");
  });

  it("prefers the label", () => {
    expect(displayName({ name: "nori-app", label: "Nori App" })).toBe("Nori App");
  });
});

File: internal/config/slug_test.go
package config

import "testing"

func TestSlug(t *testing.T) {
	tests := []struct{ in, want string }{
		{"Nori App", "nori-app"},
		{"nori-app", "nori-app"},
		{"  Nori   App  ", "nori-app"},
		{"Okane", "okane"},
		{"my_project.v2", "my_project.v2"},
		{"Ünïcødé Ãpp", "n-c-d-pp"}, // non-ASCII collapses; never leaks into a path
		{"日本語", ""},                // ...and an all-non-ASCII label yields no id at all
		{"a/b", "a-b"},              // a separator can never survive: it would escape the segment
		{"../etc", "etc"},
		{"...", ""},
		{"", ""},
		{"   ", ""},
		{"---", ""},
	}
	for _, tc := range tests {
		if got := Slug(tc.in); got != tc.want {
			t.Errorf("Slug(%q) = %q, want %q", tc.in, got, tc.want)
		}
	}
}

// Every non-empty Slug output must be a valid, stable id: slugging it again
// changes nothing, and IsSlug accepts it.
func TestSlugIsIdempotentAndSelfConsistent(t *testing.T) {
	for _, in := range []string{"Nori App", "a/b", "../etc", "my_project.v2", "Okane"} {
		s := Slug(in)
		if s == "" {
			continue
		}
		if again := Slug(s); again != s {
			t.Errorf("Slug(Slug(%q)) = %q, want %q", in, again, s)
		}
		if !IsSlug(s) {
			t.Errorf("IsSlug(Slug(%q)) = false for %q", in, s)
		}
	}
}

func TestIsSlugRejectsNonCanonical(t *testing.T) {
	for _, in := range []string{"", "Okane", "Nori App", "a/b", "-lead", "trail-", "..", "."} {
		if IsSlug(in) {
			t.Errorf("IsSlug(%q) = true, want false", in)
		}
	}
}

// SlugTyping must NOT trim the edges: trimming mid-typing makes a hyphenated id
// impossible to enter, because the separator vanishes the moment it is typed.
func TestSlugTypingKeepsTrailingSeparator(t *testing.T) {
	if got := SlugTyping("nori-"); got != "nori-" {
		t.Errorf("SlugTyping(%q) = %q, want the trailing hyphen kept", "nori-", got)
	}
	if got := SlugTyping("Nori A"); got != "nori-a" {
		t.Errorf("SlugTyping(%q) = %q, want nori-a", "Nori A", got)
	}
	// Typing a hyphenated name one keystroke at a time must converge on the slug.
	typed := ""
	for _, r := range "Nori App" {
		typed = SlugTyping(typed + string(r))
	}
	if typed != "nori-app" {
		t.Errorf("incremental typing produced %q, want nori-app", typed)
	}
}

func TestDisplayNameFallsBackToName(t *testing.T) {
	if got := (Project{Name: "nori-app"}).DisplayName(); got != "nori-app" {
		t.Errorf("DisplayName with no label = %q, want the name", got)
	}
	if got := (Project{Name: "nori-app", Label: "Nori App"}).DisplayName(); got != "Nori App" {
		t.Errorf("DisplayName = %q, want the label", got)
	}
	// A whitespace-only label is not a label.
	if got := (Project{Name: "nori-app", Label: "   "}).DisplayName(); got != "nori-app" {
		t.Errorf("DisplayName with a blank label = %q, want the name", got)
	}
}

func TestDisplayNameForUnknownProject(t *testing.T) {
	c := &Config{Projects: []Project{{Name: "nori-app", Label: "Nori App"}}}
	if got := c.DisplayNameFor("nori-app"); got != "Nori App" {
		t.Errorf("DisplayNameFor = %q, want Nori App", got)
	}
	// A session whose project was removed from config still renders something.
	if got := c.DisplayNameFor("gone"); got != "gone" {
		t.Errorf("DisplayNameFor(unknown) = %q, want the id back", got)
	}
}
